File: fifo_page_replacement.dart
import 'dart:collection';
import 'dart:io';

void main() {
  // Input: Page reference string
  stdout.write('Enter the number of pages: ');
  int n = int.parse(stdin.readLineSync()!);

  List<int> pages = [];
  stdout.write('Enter the page reference string (space-separated): ');
  pages = stdin.readLineSync()!.split(' ').map(int.parse).toList();

  if (pages.length != n) {
    print('Number of pages does not match input length!');
    return;
  }

  // Input: Number of page frames
  stdout.write('Enter the number of frames: ');
  int frameCount = int.parse(stdin.readLineSync()!);

  Queue<int> frames = Queue<int>();
  Set<int> inFrames = {};
  int pageFaults = 0;

  print('\nFIFO Page Replacement Simulation:\n');

  for (int page in pages) {
    stdout.write('Page $page -> ');

    if (!inFrames.contains(page)) {
      // Page fault
      pageFaults++;

      if (frames.length == frameCount) {
        // Remove the oldest page
        int removed = frames.removeFirst();
        inFrames.remove(removed);
      }

      frames.addLast(page);
      inFrames.add(page);
      print('Page Fault');
    } else {
      print('Hit');
    }

    print('Current Frames: ${frames.toList()}');
  }

  print('\nTotal Page Faults: $pageFaults');
}

File: fifo_page_replacement.py
from collections import deque

def fifo_page_replacement(pages, frame_count):
    frames = deque()
    in_frames = set()
    page_faults = 0

    print("\nFIFO Page Replacement Simulation:\n")

    for page in pages:
        print(f"Page {page} -> ", end='')

        if page not in in_frames:
            page_faults += 1
            if len(frames) == frame_count:
                removed = frames.popleft()
                in_frames.remove(removed)
            frames.append(page)
            in_frames.add(page)
            print("Page Fault")
        else:
            print("Hit")

        print(f"Current Frames: {list(frames)}")

    print(f"\nTotal Page Faults: {page_faults}")


if __name__ == "__main__":
    n = int(input("Enter the number of pages: "))
    pages = list(map(int, input("Enter the page reference string (space-separated): ").split()))
    
    if len(pages) != n:
        print("Number of pages does not match input length!")
    else:
        frame_count = int(input("Enter the number of frames: "))
        fifo_page_replacement(pages, frame_count)
